From e98c23238c38deb1142c299f4b00a6516180f543 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Tue, 29 Jul 2014 18:34:49 -0400
Subject: [lib] percpu_counter: unbreak __percpu_counter_add()

Message-id: <1406658889-9265-3-git-send-email-sandeen@redhat.com>
Patchwork-id: 86824
O-Subject: [RHEL7.1 PATCH 2/2] percpu_counter: unbreak __percpu_counter_add()
Bugzilla: 1123968
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Rik van Riel <riel@redhat.com>
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Hugh Dickins <hughd@google.com>

commit d1969a84dd6a44d375aa82bba7d6c38713a429c3
Author: Hugh Dickins <hughd@google.com>
Date:   Thu Jan 16 15:26:48 2014 -0800

    percpu_counter: unbreak __percpu_counter_add()

    Commit 74e72f894d56 ("lib/percpu_counter.c: fix __percpu_counter_add()")
    looked very plausible, but its arithmetic was badly wrong: obvious once
    you see the fix, but maddening to get there from the weird tmpfs ENOSPCs

    Signed-off-by: Hugh Dickins <hughd@google.com>
    Cc: Ming Lei <tom.leiming@gmail.com>
    Cc: Paul Gortmaker <paul.gortmaker@windriver.com>
    Cc: Shaohua Li <shli@fusionio.com>
    Cc: Jens Axboe <axboe@kernel.dk>
    Cc: Fan Du <fan.du@windriver.com>
    Cc: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/lib/percpu_counter.c b/lib/percpu_counter.c
index 1da85bb..8280a5d 100644
--- a/lib/percpu_counter.c
+++ b/lib/percpu_counter.c
@@ -82,7 +82,7 @@ void __percpu_counter_add(struct percpu_counter *fbc, s64 amount, s32 batch)
   unsigned long flags;
   raw_spin_lock_irqsave(&fbc->lock, flags);
   fbc->count += count;
-   __this_cpu_sub(*fbc->counters, count);
+  __this_cpu_sub(*fbc->counters, count - amount);
   raw_spin_unlock_irqrestore(&fbc->lock, flags);
  } else {
   this_cpu_add(*fbc->counters, amount);
-- 
1.7.1